internal/component/azure/servicebus/subscription.go

```diff
@@ -53,78 +62,105 @@ type Subscription struct {
+		if *opts.MaxBulkSubCount < 1 {
 			logger.Warnf("maxBulkSubCount must be greater than 0, setting it to 1")
-			maxBulkSubCount = ptr.Of(1)
+			opts.MaxBulkSubCount = ptr.Of(1)
 		}
 	} else {
 		// for non-bulk subscriptions, we only get one message at a time
-		maxBulkSubCount = ptr.Of(1)
+		opts.MaxBulkSubCount = ptr.Of(1)
 	}
 
-	if *maxBulkSubCount > maxActiveMessages {
-		logger.Warnf("maxBulkSubCount must not be greater than maxActiveMessages, setting it to %d", maxActiveMessages)
-		maxBulkSubCount = &maxActiveMessages
+	if *opts.MaxBulkSubCount > opts.MaxActiveMessages {
+		logger.Warnf("maxBulkSubCount must not be greater than maxActiveMessages, setting it to %d", opts.MaxActiveMessages)
+		opts.MaxBulkSubCount = &opts.MaxActiveMessages
 	}
 
 	s := &Subscription{
-		entity:          entity,
+		entity:          opts.Entity,
 		activeMessages:  make(map[int64]*azservicebus.ReceivedMessage),
-		timeout:         time.Duration(timeoutInSec) * time.Second,
-		maxBulkSubCount: *maxBulkSubCount,
+		timeout:         time.Duration(opts.TimeoutInSec) * time.Second,
+		maxBulkSubCount: *opts.MaxBulkSubCount,
+		requireSessions: opts.RequireSessions,
 		logger:          logger,
 		ctx:             ctx,
 		cancel:          cancel,
 		// This is a pessimistic estimate of the number of total operations that can be active at any given time.
 		// In case of a non-bulk subscription, one operation is one message.
-		activeOperationsChan: make(chan struct{}, maxActiveMessages/(*maxBulkSubCount)),
+		activeOperationsChan: make(chan struct{}, opts.MaxActiveMessages/(*opts.MaxBulkSubCount)),
 	}
 
-	if maxRetriableEPS > 0 {
-		s.retriableErrLimit = ratelimit.New(maxRetriableEPS)
+	if opts.MaxRetriableEPS > 0 {
+		s.retriableErrLimit = ratelimit.New(opts.MaxRetriableEPS)
 	} else {
 		s.retriableErrLimit = ratelimit.NewUnlimited()
 	}
 
-	if maxConcurrentHandlers > 0 {
-		s.logger.Debugf("Subscription to %s is limited to %d message handler(s)", entity, maxConcurrentHandlers)
-		s.handleChan = make(chan struct{}, maxConcurrentHandlers)
+	if opts.MaxConcurrentHandlers > 0 {
+		s.logger.Debugf("Subscription to %s is limited to %d message handler(s)", opts.Entity, opts.MaxConcurrentHandlers)
+		s.handleChan = make(chan struct{}, opts.MaxConcurrentHandlers)
 	}
 
 	return s
 }
 
 // Connect to a Service Bus topic or queue, blocking until it succeeds; it can retry forever (until the context is canceled).
-func (s *Subscription) Connect(newReceiverFunc func() (*azservicebus.Receiver, error)) error {
+func (s *Subscription) Connect(newReceiverFunc func() (Receiver, error)) (Receiver, error) {
 	// Connections need to retry forever with a maximum backoff of 5 minutes and exponential scaling.
 	config := retry.DefaultConfig()
 	config.Policy = retry.PolicyExponential
 	config.MaxInterval = 5 * time.Minute
 	config.MaxElapsedTime = 0
 	backoff := config.NewBackOffWithContext(s.ctx)
 
+	var receiver Receiver
+
 	err := retry.NotifyRecover(
```

Comment:
Suggest using [`NotifyRecoverWithData`](https://pkg.go.dev/github.com/dapr/kit/retry#NotifyRecoverWithData) to return `receiver` so there are no side effects. Although doesn't look like this method uses a goroutine inside, and so it shouldn't have risks of race conditions, avoiding side effects may be safer.

Reply:
Fixed